### ISSUE TITLE
CS_VALUE_TYPE affects parent type if applied to inner class

### ISSUE THREAD

**Issue by vonzepp** [mono/CppSharp#559]
I expect that Record remain as 'class' in generated stub and only ItemsIterator becomes struct in code below:
```
class Record
{
	CS_VALUE_TYPE class ItemsIterator
	{
	};
};
```
But both classes becomes 'struct' in generated stub.
```
        public unsafe partial struct Record
        {
...
...
            internal unsafe partial struct ItemsIterator
            {
```
Of cource I can do it by other way
```
ctx.SetClassAsValueType("ItemsIterator");
```
but default behaviour of CS_VALUE_TYPE seems to be strange.

**Comment by tritao** [comment]
Looking at the handling code in https://github.com/mono/CppSharp/blob/f97af2608716478cac4546ba3acd6ed87549f034/src/Generator/Passes/CheckMacrosPass.cs#L110 I think this behaviour can only be explained by a bug in the parser which makes both classes have the preprocessed entity.

Could you include a minified reproduction so we can fix it?

**Comment by vonzepp** [comment]
Yes, sure. But sorry I don't understand what do you mean when say
> minified reproduction
? Short project containing file.h ?:
```
#define CS_VALUE_TYPE
class Record
{
    CS_VALUE_TYPE class ItemsIterator
    {
    };
};
```

**Comment by tritao** [comment]
If that's enough to reproduce it then it's perfect, thanks.

**Comment by vonzepp** [comment]
https://www.dropbox.com/s/7wo5ruqc73j2h2q/Generator.zip?dl=0
Test project for .NET 4.6 / x64 / VS2015
```
> cd bin
> Generator.exe . .
```
First argument of generator.exe is Directory with TestFile.h, second is OutputDirectory

**Comment by tritao** [comment]
Thanks. There should no need for the binary project though, the source code above should be enough.

**Comment by genuinelucifer** [comment]
@tritao I would like to work on this (if you haven't started this yet)... I would need some help as I had a look and couldn't find how ```PreprocessedEntities``` are actually filled in any ```Class```.

**Comment by tritao** [comment]
Sure Abhinav.

Preprocessed entities are handled by the parser in Parser::HandlePreprocessedEntities, https://github.com/mono/CppSharp/blob/master/src%2FCppParser%2FParser.cpp#L2603.

I am assuming the error is in the parser, but I haven't confirmed yet. You will want to test in Debug mode to be able to properly debug the native C++ code. It can be tricky to figure out issues in the native code because Clang types can be very opaque to the debugger, so ask me for help if you get stuck.

**Comment by genuinelucifer** [comment]
Sure.
Last time i tried, i failed to compile CppSharp in Debug mode. I'll try
again and ask you if i get stuck.
Thanks :)

**Comment by genuinelucifer** [comment]
@tritao I tried debugging the code but couldn't. I also tried building C++# in debug mode but got to many clang errors and the build was never successful. Please help.

**Comment by tritao** [comment]
@genuinelucifer Feel free to try this one again now.
